python/task/workflow.py

```diff
@@ -0,0 +1,703 @@
+                format=None):
+    """Map input files, shuffle, sort, reduce, and output to files."""
+    # Determine the number of output shards.
+    shards = length_of(output)
+
+    # Mapping of input.
+    mapping = self.map(input, mapper, params=params, format=format)
+
+    # Shuffling of map output.
+    shuffle = self.shuffle(mapping, shards=shards)
+
+    # Reduction of shuffled map output.
+    self.reduce(shuffle, output, reducer, params=params)
+    return output
+
+  def start(self):
+    """Start workflow. The workflow with be run in the background, and the
+    done() and wait() methods can be used to determine if the workflow has
+    completed."""
+    # Make sure all output directories exist.
+    self.create_output_directories()
+
+    # Create underlying job in task system.
+    if self.job != None: raise Exception("job already running")
+    self.job = api.Job(self, self.name)
+
+    # Start job.
+    global active
+    active = True
+    self.job.start()
+
+  def wait(self, timeout=None):
+    """Wait until workflow completes."""
+    if self.job == None: return True
+    if timeout != None:
+      return self.job.wait_for(timeout)
+    else:
+      self.job.wait()
+      return True
+
+  def done(self):
+    """Check if workflow is done."""
+    if self.job == None: return Trye
+    return self.job.done()
+
+  def counters(self):
+    """Return map of counters for the workflow."""
+    if self.job == None: return None
+    return self.job.counters()
+
+  def dump(self):
+    """Return workflow configuration."""
+    s = ""
+    for task in self.tasks:
+      s += "task " + task.name
+      if task.shard: s += str(task.shard)
+      s += " : " + task.type
+      s += "\n"
+      for param in task.params:
+        s += "  param " + param + " = " + task.params[param] + "\n"
+      for input in task.inputs:
+        s += "  input " + str(input) + "\n"
+      for source in task.sources:
+        s += "  source " +  str(source) + "\n"
+      for output in task.outputs:
+        s += "  output " +  str(output) + "\n"
+      for sink in task.sinks:
+        s += "  sink " +  str(sink) + "\n"
+    for channel in self.channels:
+      s += "channel " + str(channel) + "\n"
+    for resource in self.resources:
+      s += "resource " + str(resource) + "\n"
+    return s
+
+  def create_output_directories(self):
+    """Create output directories for workflow."""
+    checked = set()
+    for task in self.tasks:
+      for output in task.outputs:
+        directory = os.path.dirname(output.resource.name)
+        if directory in checked: continue
+        if not os.path.exists(directory): os.makedirs(directory)
+        checked.add(directory)
+
+def start_monitor(port):
+  """Start task monitor."""
+  api.start_task_monitor(port)
+
+def stop_monitor():
+  """Stop task monitor."""
+  global active
+  if active:
+    log.info("sending final status to monitor")
+    api.finalize_dashboard();
```

Comment:
should we also set active = False here?
Also, there is a trailing semi-colon here.

Reply:
The global active flag just tracks if the workflow system has been activated, not whether it is still active. I added this so that python/run.py --help does not try to stop the monitor and write a workflow log.

Semi-colon removed.